webdriver/support/fixtures.py

```diff
@@ -115,3 +117,83 @@ def inner(path, query="", fragment=""):
+                                     "execute/sync",
+                                     dict(script="window.open();", args=[]))
+        with_new = session.send_session_command("GET", "window/handles")
+
+        assert len(initial) == len(with_new) - 1
+
+        new_handle = (set(with_new) - set(initial)).pop()
+
+        session.send_session_command("POST",
+                                     "window",
+                                     dict(handle=new_handle))
+
+        session.send_session_command("DELETE", "window")
+
+    return switch_to_inactive
+
+def create_dialog(session):
+    """Create a dialog (one of "alert", "prompt", or "confirm") and provide a
+    function to validate that the dialog has been "handled" (either accepted or
+    dismissed) by returning some value."""
+
+    def create_dialog(dialog_type):
+        dialog_id = str(uuid4())
+        dialog_text = "text " + dialog_id
+
+        assert dialog_type in ("alert", "confirm", "prompt"), (
+               "Invalid dialog type: '%s'" % dialog_type)
+
+        # Script completion and modal summoning are scheduled on two separate
+        # turns of the event loop to ensure that both occur regardless of how
+        # the user agent manages script execution.
+        spawn = """
+            var done = arguments[0];
+            window.__WEBDRIVER = "initial value {0}";
+            setTimeout(function() {{
+                done();
+            }}, 0);
+            setTimeout(function() {{
+                window.__WEBDRIVER = window.{1}("{2}");
+            }}, 0);
+        """.format(dialog_id, dialog_type, dialog_text)
+
+        session.send_session_command("POST",
+                                     "execute/async",
+                                     dict(script=spawn, args=[]))
+
+        def assert_dialog_handled(expected_value):
+            result = session.transport.send("GET",
+                                            "session/%s/alert/text" % session.session_id)
+
+            # If there were any existing dialogs prior to the creation of this
+            # fixture's dialog, then the "Get Alert Text" command will return
+            # successfully. In that case, the text must be different than that
+            # of this fixture's dialog.
+            try:
+                assert_error(result, "no such alert")
+            except:
+                assert (result.status == 200 and
+                        result.body["value"] != dialog_text), (
+                       "%s dialog was not handled." % dialog_type)
+
+            probe = "return window.__WEBDRIVER;"
+            result = session.send_session_command("POST",
+                                                  "execute/sync",
+                                                  dict(script=probe, args=[]))
+
+            assert result == expected_value, (
+                   "%s dialog was not handled with expected value" % s)
+
+        return assert_dialog_handled
```

Comment:
There is a lot of indirection and going on in this assert, and many anonymous functions nested within one another. Can you do anything about that? I’m not sure it makes sense to return an assert function from a function to create a dialogue…